### PR TITLE
Task 1: Tainted node by condition.

### LINKED ISSUE #42001: NodeController should add NoSchedule taints and we should get rid of getNodeConditionPredicate()
getNodeConditionPredicate() in plugin/pkg/scheduler/factory/factory.go makes our code hard to understand because it hides the node-condition-based filtering in the node lister, which is totally non-obvious.

We should get rid of this function and have NodeController add NoSchedule taints for these situations instead. (Alas, we might not actually be able to get rid of getNodeConditionPredicate() completely until we get rid of the Unschedulable field of PodSpec, but at least we can get rid of all the other code in this function.) If there are pods that should still be able to schedule in any of these situations (e.g. DaemonSet pods?) we should add tolerations in admission control for them (e.g. see pkg/controller/daemon/daemoncontroller.go).

cc/ @kubernetes/sig-scheduling-misc  @kubernetes/sig-cluster-lifecycle-misc 
cc/ @gmarek @kevin-wangzefeng 

**Sub-tasks according to the design doc:**

- [x] Add node taints label and feature flag (#49547)
- [x] In the node controller, taint Nodes according to the Node Condition (#49257)
- [x] In DaemonSet, update DaemonSetController to Tolerant new Taints (#50186)
- [x] In admissionController, add MemoryPressure/DiskPressure toleration for no BestEffort pod (#50180)
- [x] In scheduler, disable the current behaviour of filtering out Nodes. Instead, pods will not be scheduled to tainted nodes if a toleration does not exist in its PodSpec (#50185)
- [x] Add e2e test for this feature (https://github.com/kubernetes/kubernetes/pull/52723)
- [x] Update doc for this feature (https://github.com/kubernetes/kubernetes.github.io/pull/5352)

### PR DESCRIPTION
**Pull request by k82cn**
**What this PR does / why we need it**:
Tainted node by condition for MemoryPressure, OutOfDisk and so on.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: part of #42001 

**Release note**:
```release-note
Tainted nodes by conditions as following:
  * 'node.kubernetes.io/network-unavailable=:NoSchedule' if NetworkUnavailable is true
  * 'node.kubernetes.io/disk-pressure=:NoSchedule' if DiskPressure is true
  * 'node.kubernetes.io/memory-pressure=:NoSchedule' if MemoryPressure is true
  * 'node.kubernetes.io/out-of-disk=:NoSchedule' if OutOfDisk is true
```

### CODE REVIEW

**Comment by k82cn** [comment]
/cc @gmarek 

This PR show the major idea on tainting Node NoSchedule by condition; is this OK for you?

Test MemoryPressure in local cluster with instrument code; I'll continue to add more unit test if it align with the code in your mind :).

**Comment by k82cn** [comment]
cc/ @resouer

**Comment by k82cn** [comment]
/retest

----
Da (Klaus), Ma (马达) | PMP® | Software Architect, IBM Cloud private
Development
IBM Spectrum Computing, IBM System
+86-10-8245 4084 | klaus1982.cn@gmail.com | http://k82.me

File: pkg/controller/controller_utils.go [hunk 1]
```diff
@@ -901,13 +905,22 @@ func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taint *v1.Ta
...
+		oldNodeCopy := oldNode
+		for _, taint := range taints {
+			curNewNode, ok, err := taintutils.AddOrUpdateTaint(oldNodeCopy, taint)
+			if err != nil {
+				return fmt.Errorf("Failed to update taint annotation!")
+			}
+			if !ok {
+				return nil
+			}
+
```
Comment (gmarek) [inline pkg/controller/controller_utils.go]: remove empty line

File: pkg/controller/controller_utils.go [hunk 2]
```diff
@@ -885,7 +885,11 @@ func (o ReplicaSetsBySizeNewer) Less(i, j int) bool {
 	return *(o[i].Spec.Replicas) > *(o[j].Spec.Replicas)
 }
 
-func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taint *v1.Taint) error {
+func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taints ...*v1.Taint) error {
```
Comment (gmarek) [inline pkg/controller/controller_utils.go]: Can you put those changes into separate PR? They have a lot of sense on their own, and we don't want to revert those in case we find a bug somewhere else.

File: pkg/controller/node/controller_utils.go [hunk 3]
```diff
@@ -290,31 +290,35 @@ func recordNodeStatusChange(recorder record.EventRecorder, node *v1.Node, new_st
...
+		taintToAdd.TimeAdded = metav1.Now()
+	}
+
+	err := controller.AddOrUpdateTaintOnNode(kubeClient, node.Name, taintsToAdd...)
 	if err != nil {
 		utilruntime.HandleError(
 			fmt.Errorf(
 				"unable to taint %v unresponsive Node %q: %v",
-				taintToAdd.Key,
+				taintsToAdd,
```
Comment (gmarek) [inline pkg/controller/node/controller_utils.go]: This will print pointers. Please make sure a sane error message is printed.

Comment (gmarek) [inline pkg/controller/node/controller_utils.go]: this will print pointer addresses. Make sure that it prints something understandable.

File: pkg/controller/node/nodecontroller.go [hunk 4]
```diff
@@ -153,8 +167,12 @@ type NodeController struct {
 	// workers that evicts pods from unresponsive nodes.
 	zonePodEvictor map[string]*RateLimitedTimedQueue
 	// workers that are responsible for tainting nodes.
-	zoneNotReadyOrUnreachableTainer map[string]*RateLimitedTimedQueue
-	podEvictionTimeout              time.Duration
+	zoneNoExecuteTainer map[string]*RateLimitedTimedQueue
+
+	noScheduleTainerLock sync.Mutex
+	zoneNoScheduleTainer map[string]*RateLimitedTimedQueue
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: I'll need to read it carefully (I'll try to do this in the evening, but I may fail, so I'm leaving this note here), but I think this will be wrong. We want to have evictor (i.e. rate-limited tainting) only for Unreachable/NotReady Taints. The rest of the taints should not need rate limiting as they're NoSchedule ones. The only thing that we should protect against is TaintFlapping (i.e. adding and removing Taints in quick succession).

As I said - it's possible that this is how it's done, but I didn't have time to read it carefully enough this time.

File: pkg/controller/controller_utils.go (hunk 2 above)
Comment (k82cn) [inline pkg/controller/controller_utils.go]: sure, created PR https://github.com/kubernetes/kubernetes/pull/49524 .

File: plugin/pkg/scheduler/algorithm/well_known_labels.go [hunk 5]
```diff
@@ -28,6 +28,26 @@ const (
 	// and removed when node becomes reachable (NodeReady status ConditionTrue).
 	TaintNodeUnreachable = "node.alpha.kubernetes.io/unreachable"
 
+	// When feature-gate for TaintBasedEvictions=true flag is enabled,
```
Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: Also create PR https://github.com/kubernetes/kubernetes/pull/49547 for this diff; if we revert taints part, it will not introduce build error.

File: pkg/controller/node/nodecontroller.go [hunk 6]
```diff
@@ -435,12 +454,58 @@ func (nc *NodeController) doEvictionPass() {
 	}
 }
 
-func (nc *NodeController) doTaintingPass() {
+func (nc *NodeController) doNoScheduleTaintingPass() {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Yeah - this is not needed, we don't need to rate limit condition-based tainting.

File: pkg/controller/node/nodecontroller.go [hunk 7]
```diff
@@ -1067,23 +1140,35 @@ func (nc *NodeController) evictPods(node *v1.Node) bool {
...
+
+	// TODO(k82cn): refactor NodeController to the following routines:
+	//   - monitorNodeStatus: update NodeConditions, limiter and cluster/zone disruption
+	//   - tainter: taints Node according to NodeCondition, e.g. checking cluster/zone disruption
+	//              before taints, did not update node if not condition changed
+	nc.zoneNoScheduleTainer[utilnode.GetZoneKey(node)].Remove(node.Name)
+	return nc.zoneNoScheduleTainer[utilnode.GetZoneKey(node)].Add(node.Name, string(node.UID))
 }
 
 func (nc *NodeController) markNodeAsHealthy(node *v1.Node) (bool, error) {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Once we introduce new Taints, this should be renamed to 'markNodeAsReachable'

File: pkg/controller/node/nodecontroller.go [hunk 8]
```diff
@@ -598,6 +667,10 @@ func (nc *NodeController) monitorNodeStatus() error {
 			zoneToNodeConditions[utilnode.GetZoneKey(node)] = append(zoneToNodeConditions[utilnode.GetZoneKey(node)], currentReadyCondition)
 		}
 
+		if nc.useTaintBasedEvictions {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: This should have it's own flag-gate, as this is different feature.

File: pkg/controller/node/nodecontroller.go [hunk 9]
```diff
@@ -598,6 +667,10 @@ func (nc *NodeController) monitorNodeStatus() error {
...
+			nc.syncupNodeForNoScheduleTainting(node)
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: We shouldn't do this in 'monitorNodeStatus', as it runs every 10s (or sth). This would mean that there can be designed 10s delay between Node getting a Condition and Taint appearance, which is not enough. For those kind of taints we just need to register a function in Informer that will fire on every Node create and update.

File: pkg/controller/node/nodecontroller.go (hunk 4 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Yup, I stand by what I wrote. We don't need rate limiting for those taints.

File: pkg/controller/node/nodecontroller.go (hunk 9 above)
Comment (k82cn) [inline pkg/controller/node/nodecontroller.go]: done.

File: pkg/controller/node/nodecontroller.go (hunk 8 above)
Comment (k82cn) [inline pkg/controller/node/nodecontroller.go]: done.

File: pkg/controller/node/nodecontroller.go (hunk 7 above)
Comment (k82cn) [inline pkg/controller/node/nodecontroller.go]: done

File: pkg/controller/node/controller_utils.go (hunk 3 above)
Comment (k82cn) [inline pkg/controller/node/controller_utils.go]: done

File: pkg/controller/node/nodecontroller.go [hunk 10]
```diff
@@ -152,8 +166,8 @@ type NodeController struct {
 	// workers that evicts pods from unresponsive nodes.
 	zonePodEvictor map[string]*RateLimitedTimedQueue
 	// workers that are responsible for tainting nodes.
-	zoneNotReadyOrUnreachableTainer map[string]*RateLimitedTimedQueue
-	podEvictionTimeout              time.Duration
+	zoneNoExecuteTainer map[string]*RateLimitedTimedQueue
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Nit - can you put this rename into separate PR - it'll make review much easier.

File: pkg/controller/node/nodecontroller.go [hunk 11]
```diff
@@ -434,12 +465,40 @@ func (nc *NodeController) doEvictionPass() {
 	}
 }
 
-func (nc *NodeController) doTaintingPass() {
+func (nc *NodeController) doNoScheduleTaintingPass(node *v1.Node) error {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Much better - thanks. I believe it's also way simpler;)

**Comment by gmarek** [comment]
It looks good to me, but I'd do more careful review after rename is moved to a separate PR.

File: pkg/controller/node/nodecontroller.go (hunk 10 above)
Comment (k82cn) [inline pkg/controller/node/nodecontroller.go]: Open #49870 for this :).

File: plugin/pkg/scheduler/algorithm/well_known_labels.go [hunk 12]
```diff
@@ -28,6 +28,26 @@ const (
...
+	// TaintNodeDiskPressure would be automatically added by node controller
+	// when node has disk pressure, and removed when node has enough disk.
+	TaintNodeDiskPressure = "node.alpha.kubernetes.io/diskPressure"
+
+	// When feature-gate for TaintBasedEvictions=true flag is enabled,
+	// TaintNodeNetworkUnavailable would be automatically added by node controller
+	// when node's network is unavailable, and removed when network becomes ready.
+	TaintNodeNetworkUnavailable = "node.alpha.kubernetes.io/networkUnavailable"
+
```
Comment (yastij) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: @k82cn - don't we need tainting based on inode pressure ?

```go
TaintNodeInodePressure = "node.alpha.kubernetes.io/InodePressure"
```

Comment (yastij) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: @k82cn @gmarek - don't we need tainting based on inode pressure ?

```go
TaintNodeInodePressure = "node.alpha.kubernetes.io/InodePressure"
```

Comment (gmarek) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: It depends what's the plan for this condition. I recall hearing that it's deprecated, but I may be wrong. @kubernetes/sig-node-pr-reviews

Comment (yastij) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: @gmarek If it's deprecated we should state it somewhere - docs / [design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/taint-node-by-condition.md) -. If not we should add it.

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: Good point :). After got sig-node's feedback, I'll align them. And after grep `NodeInodePressure`, it seems kubelet did not use it anymore.

Comment (yastij) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: LGTM then

Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/well_known_labels.go]: here's the background (https://github.com/kubernetes/kubernetes/pull/50124) :). I'll update the design doc when that PR merged.

**Comment by k82cn** [comment]
rebased :).

**Comment by gmarek** [comment]
Aaand another rebase is needed.

File: pkg/controller/node/nodecontroller_test.go [hunk 13]
```diff
@@ -2060,6 +2062,186 @@ func TestSwapUnreachableNotReadyTaints(t *testing.T) {
 	}
 }
 
+func TestTaintsNodeByCondition(t *testing.T) {
+	fakeNow := metav1.Date(2017, 1, 1, 12, 0, 0, 0, time.UTC)
+	evictionTimeout := 10 * time.Minute
+
+	fakeNodeHandler := &testutil.FakeNodeHandler{
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller_test.go]: Why not do this similarly to other tests and define it inline in test case definition?

**Comment by gmarek** [comment]
One very minor nit. LGTM otherwise.

@davidopp - this is the main PR in this effort. What's left is to change all the other places in control plane that are using Conditions to consider Taints instead. PRs are mostly in flight, so it's rather safe to say that this will land for 1.8.

@k82cn - one thing that would be awesome is to have e2e test for tainting/admission controller/etc. It certainly will be needed for beta, and I'd have more confidence in the fact that we haven't missed anything if it was written for alpha as well. Luckily there's not a lot of time pressure on it, as we can add test during stabilization phase:)

**Comment by k82cn** [comment]
regarding e2e, yes, it's one of sub-task here (https://github.com/kubernetes/kubernetes/issues/42001#issuecomment-315987635); I'll add e2e when all related PR merged :).

**Comment by gmarek** [comment]
You'll need help from someone from google to start running the test continuously, but that can be arranged:)

File: pkg/controller/node/nodecontroller_test.go (hunk 13 above)
Comment (k82cn) [inline pkg/controller/node/nodecontroller_test.go]: In this test, it only care about the update of one node status; it's not necessary to build `fakeNodeHandler` with different existing nodes :).

**Comment by k82cn** [comment]
@davidopp , @wojtek-t , do you have any comments for this :).

**Comment by davidopp** [comment]
I will take a look this week. May not be able to do it for a few days.

**Comment by k82cn** [comment]
ping :).

**Comment by wojtek-t** [comment]
/approve

I will let @davidopp LGTM (it looks good to me).

**Comment by luxas** [comment]
Just noting that `networkUnavailable` should be `network-unavailable` before moving to beta...
Is that something we can do at this stage?
(I know this goes for the general TaintBasedEvictions feature, which still is alpha)

cc @thockin

**Review by resouer — COMMENTED** [review]
@luxas This PR does not rely on well known labels, here is a separate fix #51266 , I am asking @k82cn to review.

**Comment by k82cn** [comment]
@luxas , as we slack, there's not hard code in this feature; https://github.com/kubernetes/kubernetes/pull/51266 is enough to cover, and no impact to this PR.

**Comment by resouer** [comment]
Cool, but I think you may still need to update the PR description.

**Comment by yastij** [comment]
We'll also need to align the design doc

**Comment by luxas** [comment]
@k82cn @resouer Yeah, cool, just stumbled upon that in the release note above; we should replace it there as well. Thanks!

**Comment by k82cn** [comment]
the description was updated :).

@davidopp , would you help to review this PR?

**Comment by gmarek** [comment]
Ping @davidopp

**Comment by gmarek** [comment]
I spoke with David offline. Applying label.
